doc/contributing: enforce full commit hashes on github (#189241)
NixOS/nixpkgs: doc/contributing/coding-conventions.chapter.md

  }
  ```

> When fetching from GitHub commits must always be referenced by their full commit hash. This is because GitHub shares commit hashes between all forks and returns a 404 if a short commit hash is ambigious. For some 6 character commits in nixpkgs this is already the case. It is a practical DoS attack to push great amounts of auto generated commits into forks and was already [demonstrated against GitHub Actions Beta](https://blog.teddykatz.com/2019/11/12/github-actions-dos.html).

Reviewer: ```suggestion
When fetching from GitHub, commits must always be referenced by their full commit hash.
This is because GitHub shares commit hashes between all forks and returns [404 Not Found][http-404] if a short commit hash is ambiguous. For some 6 character commits in `nixpkgs` this is already the case.

It is a practical DoS attack to push great amounts of auto generated commits into forks and was already [demonstrated against GitHub Actions Beta][github-actions-dos].

[http-404]: https://www.rfc-editor.org/rfc/rfc7231#section-6.5.4
[github-actions-dos]: https://blog.teddykatz.com/2019/11/12/github-actions-dos.html
```

Author: Added the grammar changes but ignored the 404 RFC link (we don't really need that here) and didn't extract the links because we don't use them in multiple places.

Reviewer: @SuperSandro2000 Okay, let me give more context for why I suggested it this way and not otherwise:

- reference links

  While this is neither accepted nor debated, I am trying to establish conventions for documentation, and provide rationale. See here specifically on [line wrapping and reference links](https://github.com/NixOS/nix.dev/pull/261/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R169-R179).

- short paragraphs

  Both based on anecdotal evidence as well as [authoritatively-sounding sources](https://www.plainlanguage.gov/guidelines/concise/write-short-paragraphs/) I insist on more white space in prose writing. Most Nix documentation is clumsily worded *and* visually dense, i.e. it's just hard to read. Naturally, most contributors *and* readers are not native English speakers - which is fine, and we have to get to better language iteratively, but it produces quite some of friction. What can do right now is slightly improve structure on every change.
  
- links to definitions

  We appear to have converged on using [culturally neutral language](https://github.com/NixOS/nix.dev/blob/master/CONTRIBUTING.md?plain=1#L17-L21), and I will probably make another update to accommodate the notion that in-group knowledge is also culture. (Here is a more [emotionally charged comment](https://github.com/NixOS/nixpkgs/pull/176557#discussion_r932742878) on the topic.) Assuming your readers know what 404 means without providing context is quite a bold move. If you don't define a technical term inline, please link to an authoritative definition.

Author: > Assuming your readers know what 404 means without providing context is quite a bold move. If you don't define a technical term inline, please link to an authoritative definition.

Sure we can improve the wording a bit and add Not Found but linking the RFC which contains the technical implementation details is not helpful. In the context it does not matter what the server returns or what it technically means, the important part is that the server does not longer return the patch and an uncached build fails.

---

Also this discussion is a bit off topic for this change. Can we move this to somewhere else?

  }
  ```

> When fetching from GitHub, commits must always be referenced by their full commit hash. This is because GitHub shares commit hashes between all forks and returns 404 if a short commit hash is ambiguous. For some 6-character commit hashes in `nixpkgs` this is already the case.

Reviewer 2: English is not my native tongue, however I suggest:
```suggestion
When fetching from GitHub, commits must always be referenced by their full commit hash. This is because GitHub shares commit hashes among all forks and returns 404 when a short commit hash is ambiguous. It already happened for some short, 6-character commit hashes in `nixpkgs`.
```
